astropy/coordinates/spectral_coordinate.py

```diff
@@ -0,0 +1,842 @@
+                              differential_type='cartesian')
+        elif isinstance(frame, str):
+            frame_cls = frame_transform_graph.lookup_name(frame)
+            frame = frame_cls(0 * u.m, 0 * u.m, 0 * u.m,
+                              0 * u.m / u.s, 0 * u.m / u.s, 0 * u.m / u.s,
+                              representation_type='cartesian',
+                              differential_type='cartesian')
+
+        observer = update_differentials_to_match(self.observer, frame,
+                                                 preserve_observer_frame=preserve_observer_frame)
+
+        new_coord = self._change_observer_to(observer)
+
+        return new_coord
+
+    def with_radial_velocity_shift(self, target_shift=None, observer_shift=None):
+        """
+        Apply a velocity shift to this spectral coordinate. The shift
+        can be provided as a redshift (float value) or radial velocity
+        (`~astropy.units.Quantity` with physical type of 'speed').
+
+        Parameters
+        ----------
+        target_shift : float or `~astropy.units.Quantity`
+            Shift value to apply to current target.
+        observer_shift : float or `~astropy.units.Quantity`
+            Shift value to apply to current observer.
+
+        Returns
+        -------
+        `SpectralCoord`
+            New spectral coordinate with the target/observer velocity changed
+            to incorporate the shift.
+        """
+        if observer_shift is not None and (self.target is None or
+                                           self.observer is None):
+            raise ValueError("Both an observer and target must be defined "
+                             "before applying a velocity shift.")
+
+        for arg in [x for x in [target_shift, observer_shift] if x is not None]:
+            if isinstance(arg, u.Quantity) and \
+                    arg.unit.physical_type not in ['speed', 'dimensionless']:
+                raise u.UnitsError("Argument must have unit physical type "
+                                   "'speed' for radial velocty or "
+                                   "'dimesionless' for redshift.")
+
+        # The target or observer value is defined but is not a quantity object,
+        #  assume it's a redshift float value and convert to velocity
+
+        if target_shift is not None:
+            target_shift = u.Quantity(target_shift)
+            if target_shift.unit.physical_type == 'dimensionless':
+                target_shift = _redshift_to_velocity(target_shift)
+
+        if observer_shift is not None:
+            observer_shift = u.Quantity(observer_shift)
+            if observer_shift.unit.physical_type == 'dimensionless':
+                observer_shift = _redshift_to_velocity(observer_shift)
+
+        target_icrs = self._target.transform_to(ICRS)
+        observer_icrs = self._observer.transform_to(ICRS)
+
+        target_shift = 0 * u.km / u.s if target_shift is None else target_shift
```

Comment:
Maybe have this as an `else` clause at the end of `if target_shift is not None` above?

Reply:
Done